### PR TITLE
944 Make CourseParticipation`setting` and `outcome` fields optional

### PR DESCRIPTION
**Pull request by nsmj63-dxw**
## Context

> [see #944 on the Accredited Programmes Trello board.](https://trello.com/c/Vtpg6D7G/944-make-setting-and-outcome-not-required-on-a-courseparticipation-make-type-status-required-api-s)

## Changes in this PR

We [recently](https://github.com/ministryofjustice/hmpps-accredited-programmes-api/pull/133) updated the API to accept an empty object when creating a `CourseParticipation`. Manually submitting its empty embedded objects on creation -`CourseParticipationSetting` and `CourseOutcome` - now seems redundant.

This ticket focuses on removing these as required fields in the OpenApi specification, when `POST`ing to `/course-participations`:

- Removing the `required` flag from `setting` on `CourseParticipation`, and giving the `type` enum property on `CourseParticipationSetting` the `required` flag instead.
- Removing the `required` flag from `outcome` on `CourseParticipation`, and giving the `status` enum property on `CourseOutcome` the `required` flag instead.

This allows us to `POST` a null `setting` or `outcome` object when necessary.

Relevant tests have also been updated (and renamed, where they don't fit the current structure).

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes UI for this change to work...
  - [ ] ... and they been released to production already

### Post-merge

- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt [hunk 1]
```diff
@@ -77,27 +76,21 @@ data class CourseParticipationSetting(
...
-  private val ignoreMe: Int = 0, // This unused, non-nullable field forces Hibernate to create an @Embedded instance when all fields are null.
+  var type: CourseSetting,
 )
 
 @Embeddable
 data class CourseOutcome(
   @Enumerated(EnumType.STRING)
   @Column(name = "outcome_status")
-  var status: CourseStatus? = null,
+  var status: CourseStatus,
```
Comment (yndajas) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt:86]: This is probably for a separate PR, but would it make sense to namespace the `CourseSetting` and `CourseStatus` enums more explicitly, e.g. `CourseParticipationSettingType` and `CourseParticipationOutcomeStatus`?

They're both about course participations, not courses, and having a `Setting` be a property of a `Setting` is maybe a little confusing?

Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt:86]: Agreed. I'd actually prefer to simplify this down even further - i.e. have a flat structure across the entire entity, and just shift things up a layer. But I don't mind sticking to making name changes here / in a follow-up PR for now.

Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt:86]: I'll implement the name changes in an immediate follow-up PR in exchange for an approval now 😁

**Review by Gweaton — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt (hunk 1 above)
Comment (yndajas) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/participationhistory/domain/CourseParticipation.kt:86]: This should also be a `CourseParticipationOutcome` I guess, but if you're flattening that might be irrelevant

**Pull request merged**